lib/AST/ASTMangler.cpp

```diff
@@ -1470,48 +1469,61 @@ void ASTMangler::appendFunctionType(AnyFunctionType *fn,
   }
 }
 
-void ASTMangler::appendFunctionSignature(AnyFunctionType *fn,
-                                         bool forceSingleParam) {
-  appendParams(fn->getResult(), /*forceSingleParam*/ false);
-  appendParams(fn->getInput(), forceSingleParam);
+void ASTMangler::appendFunctionSignature(AnyFunctionType *fn) {
+  appendFunctionResultType(fn->getResult());
+  appendFunctionInputType(fn->getParams());
   if (fn->throws())
     appendOperator("K");
 }
 
-void ASTMangler::appendParams(Type ParamsTy, bool forceSingleParam) {
-  if (TupleType *Tuple = ParamsTy->getAs<TupleType>()) {
-    if (Tuple->getNumElements() == 0) {
-      if (forceSingleParam) {
-        // A tuple containing a single empty tuple.
+void ASTMangler::appendFunctionInputType(
+    ArrayRef<AnyFunctionType::Param> params) {
+  switch (params.size()) {
+  case 0:
+    appendOperator("y");
+    break;
+
+  case 1: {
+    auto &param = params[0];
+    if (auto *tupleType = param.getType()->getAs<TupleType>()) {
+      // A paren containing a single empty tuple.
+      if (tupleType->getNumElements() == 0) {
         appendOperator("y");
         appendOperator("t");
         appendListSeparator();
         appendOperator("t");
-      } else {
-        appendOperator("y");
-      }
-      return;
-    }
-    if (forceSingleParam && Tuple->getNumElements() > 1) {
-      auto flags = ParameterTypeFlags();
-      if (ParenType *Paren = dyn_cast<ParenType>(ParamsTy.getPointer())) {
-        ParamsTy = Paren->getUnderlyingType();
-        flags = Paren->getParameterFlags();
+        return;
       }
 
-      appendType(ParamsTy);
-      if (flags.isShared())
+      appendType(param.getType());
+      if (param.isShared())
         appendOperator("h");
       appendListSeparator();
       appendOperator("t");
       return;
     }
+
+    appendType(param.getType());
+    break;
   }
 
-  if (ParenType *Paren = dyn_cast<ParenType>(ParamsTy.getPointer()))
-    ParamsTy = Paren->getUnderlyingType();
+  default: {
+    bool firstParameter = true;
+    for (auto &param : params) {
+      appendTypeListElement(param.getLabel(), param.getType(),
+                            param.getParameterFlags());
+      appendListSeparator(firstParameter);
+    }
+    break;
+  }
+  }
+}
+
+void ASTMangler::appendFunctionResultType(Type ResultTy) {
+  if (ParenType *Paren = dyn_cast<ParenType>(ResultTy.getPointer()))
+    ResultTy = Paren->getUnderlyingType();
```

Comment:
We shouldn't need to do this, right? The result type won't be arbitrarily parenthesized.

Reply:
Since this is not canonical form result type might still have parens which we have to look through e.g. `() -> ((Int, Int))`

Reply:
But we are working with canonical forms here, generally, aren't we? The result type doesn't maintain parentheses in its canonical form.

Reply:
Yeah, mangling only mangles canonical types, except for in the special case of DWARF debug info, but there we don't need to maintain 100% fidelity with parens (but preserving typealias sugar is generally desirable).